### ISSUE TITLE
Spurious "A type predicate's type must be assignable to its parameter's type"

### ISSUE THREAD

**Issue by pelotom** [Microsoft/TypeScript#21185]
**TypeScript Version:**  2.7.0-dev.20180113

**Code**

This type operator from the [Unionize](https://github.com/pelotom/unionize) library:

```ts
type Predicates<TaggedRecord> = {
  [T in keyof TaggedRecord]: (variant: TaggedRecord[keyof TaggedRecord]) => variant is TaggedRecord[T]
}
```

**Expected behavior:**

Should compile, and does in `typescript@2.6.2`.

**Actual behavior:**

Fails to compile in `typescript@next` with

```
A type predicate's type must be assignable to its parameter's type.
  Type 'TaggedRecord[T]' is not assignable to type 'TaggedRecord[keyof TaggedRecord]'.
```

**Comment by sandersn** [comment]
Caused by #17912. Before that, the assignability was not checked at all; the constraint of both types was incorrectly `any`. Now, neither type has a constraint, so they are incorrectly not assignable. Turns out our assignability rules for indexed access types are incomplete. 

I'll put up a PR after lunch; I need to write some more tests.

**Comment by sandersn** [comment]
Fix is up at #21242
